### PR TITLE
Adding lib for exposing rootCA

### PR DESCRIPTION
**Pull request by Gmerold**
# Description

Adds a library to `magma-orc8r-certifier` to expose `rootCA`.
This PR is part of [TELCO-184](https://warthogs.atlassian.net/jira/software/c/projects/TELCO/boards/51?selectedIssue=TELCO-184). 

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+```shell
+charmcraft fetch-lib charms.magma_orc8r_certifier.v0.cert_root
+```
+
+
+### Requirer Charm
+
+```python
+class CertRootRequires(Object):
```
Comment (saltiyazan) [inline orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root.py]: A provider example is missing :)

File: orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    on = CertRootRequirerCharmEvents()
+
+    def __init__(self, charm, relationship_name: str):
+        self.relationship_name = relationship_name
+        self.charm = charm
+        super().__init__(charm, relationship_name)
+        self.framework.observe(
+            charm.on[relationship_name].relation_joined, self._on_relation_changed
```
Comment (saltiyazan) [inline orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root.py]: shouldn't is be observing `CertRootRequirerCharmEvents` such as `certificate_available`?

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_cert_root.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    def test_given_cert_root_relation_when_set_certificate_then_certificate_is_added_to_relation_data(  # noqa: E501
+        self,
+    ):
+        certificate = "whatever cert"
+        relation_id = self.harness.add_relation(
+            relation_name=self.relationship_name, remote_app="whatever app"
+        )
+
+        self.harness.charm.root_cert.set_certificate(
```
Comment (saltiyazan) [inline orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_cert_root.py]: I'd say this should be part of the charm itself, and be called when handling `certificate_request` event

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+from ops.main import main
+
+logger = logging.getLogger(__name__)
+
+
+class WhateverCharm(CharmBase):
+    def __init__(self, *args):
+        """Creates a new instance of this object for each event."""
+        super().__init__(*args)
+        self.root_cert = CertRootProvides(self, "cert-root")
```
Comment (saltiyazan) [inline orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py]: We could observe `certificate_request` and set the certificate.

File: orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root.py (hunk 1 above)
Comment (Gmerold) [reply orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root.py]: Docstrings improved.

File: orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root.py (hunk 2 above)
Comment (Gmerold) [reply orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root.py]: Agreed. Fixed.

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_cert_root.py (hunk 3 above)
Comment (Gmerold) [reply orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_cert_root.py]: You're right. Corrected.

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py (hunk 4 above)
Comment (Gmerold) [reply orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py]: Agreed. Fixed.

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        """Creates a new instance of this object for each event."""
+        super().__init__(*args)
+        self.root_cert_provider = CertRootCAProvides(self, "cert-root")
+
+        self.framework.observe(
+            self.root_cert_provider.on.certificate_request, self._on_certificate_request
+        )
+
+    def _on_certificate_request(self, event):
+        logger.error("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
```
Comment (saltiyazan) [inline orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py]: Is this leftover from debugging :D ?

Comment (Gmerold) [reply orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py]: Yes :D Removed.

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import logging
+
+from charms.magma_orc8r_certifier.v0.cert_root import CertRootCAProvides
+from ops.charm import CharmBase
+from ops.main import main
+
+logger = logging.getLogger(__name__)
+
+
+class WhateverCharm(CharmBase):
```
Comment (saltiyazan) [inline orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py]: Do we need also a `requirer` dummy charm for testing the other side? or is it enough as is now?

Comment (Gmerold) [reply orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_provider/src/charm.py]: Well, I wanted to make it fast and minimal, but you're right, let's not cut corners. 
Added ;)

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_ca_requirer/src/charm.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env python3
```
Comment (sanchezfdezjavier) [inline orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_ca_requirer/src/charm.py:1]: Do we need this line?

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_ca_provider/src/charm.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python3
```
Comment (sanchezfdezjavier) [inline orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_ca_provider/src/charm.py]: Do we need this line?

**Comment by saltiyazan** [comment]
Looks good to me now :)
One general question I still have, if the library should care about the events being emitted by a `leader` unit?
or are you leaving it to the implementing charm to decide?

**Review by saltiyazan — APPROVED** [review]

File: orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root_ca.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        Args:
+            event (RelationChangedEvent): Juju event
+
+        Returns:
+            None
+        """
+        relation_data = event.relation.data
+        certificate = relation_data[event.unit].get("certificate")
+        if certificate:
+            self.on.certificate_available.emit(certificate=certificate)
```
Comment (sanchezfdezjavier) [inline orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root_ca.py:225]:
```suggestion
        if certificate := relation_data[event.unit].get("certificate"):
            self.on.certificate_available.emit(certificate=certificate)
```

**Comment by Gmerold** [comment]
> Looks good to me now :) One general question I still have, if the library should care about the events being emitted by a `leader` unit? or are you leaving it to the implementing charm to decide?

It doesn't really matter. When `Provider` pushes the cert to the relation data bag, `Requirer` sees the `relation_changed` event. Using `event.unit` it knows exactly which unit has the data, fetches it and emits `certificate_available` with the cert inside the event.

**Review by sanchezfdezjavier — CHANGES_REQUESTED** [review]

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_ca_requirer/src/charm.py (hunk 7 above)
Comment (Gmerold) [reply orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_ca_requirer/src/charm.py:1]: We don't need need it. I mean charm will work without it, but then again we keep it in all of our charms.

File: orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_ca_provider/src/charm.py (hunk 8 above)
Comment (Gmerold) [reply orchestrator-bundle/orc8r-certifier-operator/tests/unit/lib/charms/magma_orc8r_certifier/v0/test_charms/test_root_ca_provider/src/charm.py]: As above.

File: orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root_ca.py (hunk 9 above)
Comment (Gmerold) [reply orchestrator-bundle/orc8r-certifier-operator/lib/charms/magma_orc8r_certifier/v0/cert_root_ca.py:225]: I guess that's just matter of personal preferences, but I think that the current version is easier to read. If you don't mind, I'd keep it as is.

**Review by sanchezfdezjavier — APPROVED** [review]

**Pull request merged**